### ISSUE TITLE
Clang -Wcast-qual Incorrectly Triggers on a Typedef Pointer Type

### ISSUE THREAD

**Issue by gavinhoward** [llvm/llvm-project#53886]
When running `-Wcast-qual` as part of `-Weverything` (I know, it's not recommended; I like it, and I turn off specific warnings), Clang incorrectly warns about casting a `const`-qualified pointer to a non-`const`-qualified pointer, if it is being cast to a type that is `typedef`'ed to a pointer type.

Minimal example:

```
#include <stdint.h>

typedef unsigned char* my_ptr1;
typedef unsigned short* my_ptr2;
typedef unsigned int* my_ptr4;
typedef unsigned long* my_ptr8;

typedef unsigned char my_int1;
typedef unsigned short my_int2;
typedef unsigned int my_int4;
typedef unsigned long my_int8;

#define MY_MEMCPY_INC(n) \
{ \
    *((my_ptr ## n) d) = *((const my_ptr ## n) s); \
    d += n; \
    s += n; \
}
void*
mymemcpy(void* restrict dest, const void* restrict src, size_t n);

void*
mymemcpy(void* restrict dest, const void* restrict src, size_t n)
{
	unsigned char* restrict d = dest;
	const unsigned char* restrict s = src;

	(void) n;

	MY_MEMCPY_INC(1)
	MY_MEMCPY_INC(2)
	MY_MEMCPY_INC(4)
	MY_MEMCPY_INC(8)

	*((my_ptr1) d) = *((const my_ptr1) s);
	*((my_ptr2) d) = *((const my_ptr2) s);
	*((my_ptr4) d) = *((const my_ptr4) s);
	*((my_ptr8) d) = *((const my_ptr8) s);

	*((unsigned char*) d) = *((const unsigned char*) s);
	*((unsigned short*) d) = *((const unsigned short*) s);
	*((unsigned int*) d) = *((const unsigned int*) s);
	*((unsigned long*) d) = *((const unsigned long*) s);

	*((my_int1*) d) = *((const my_int1*) s);
	*((my_int2*) d) = *((const my_int2*) s);
	*((my_int4*) d) = *((const my_int4*) s);
	*((my_int8*) d) = *((const my_int8*) s);

	return d;
}
```

(I know this code is incorrect for a `memcpy()`; the original code was a `memcpy()` clone, and I reduced its size.)

If you save the above example as `test.c`, I also have a `Makefile` for convenience:

```
test:
	clang -Werror -Weverything -Wno-cast-align -o test.o -c test.c

clean:
	rm -rf test.o
```

Output of `clang --version`:

```
clang version 13.0.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm/13/bin
```

It is the Clang packaged by Gentoo.

If you run `make`, the output is as follows:

```
clang -Werror -Weverything -Wno-cast-align -o test.o -c test.c
test.c:30:2: error: cast from 'const unsigned char *' to 'unsigned char *' drops const qualifier [-Werror,-Wcast-qual]
        MY_MEMCPY_INC(1)
        ^
test.c:15:45: note: expanded from macro 'MY_MEMCPY_INC'
        *((my_ptr ## n) d) = *((const my_ptr ## n) s); \
                                                   ^
test.c:31:2: error: cast from 'const unsigned char *' to 'unsigned short *' drops const qualifier [-Werror,-Wcast-qual]
        MY_MEMCPY_INC(2)
        ^
test.c:15:45: note: expanded from macro 'MY_MEMCPY_INC'
        *((my_ptr ## n) d) = *((const my_ptr ## n) s); \
                                                   ^
test.c:32:2: error: cast from 'const unsigned char *' to 'unsigned int *' drops const qualifier [-Werror,-Wcast-qual]
        MY_MEMCPY_INC(4)
        ^
test.c:15:45: note: expanded from macro 'MY_MEMCPY_INC'
        *((my_ptr ## n) d) = *((const my_ptr ## n) s); \
                                                   ^
test.c:33:2: error: cast from 'const unsigned char *' to 'unsigned long *' drops const qualifier [-Werror,-Wcast-qual]
        MY_MEMCPY_INC(8)
        ^
test.c:15:45: note: expanded from macro 'MY_MEMCPY_INC'
        *((my_ptr ## n) d) = *((const my_ptr ## n) s); \
                                                   ^
test.c:35:37: error: cast from 'const unsigned char *' to 'unsigned char *' drops const qualifier [-Werror,-Wcast-qual]
        *((my_ptr1) d) = *((const my_ptr1) s);
                                           ^
test.c:36:37: error: cast from 'const unsigned char *' to 'unsigned short *' drops const qualifier [-Werror,-Wcast-qual]
        *((my_ptr2) d) = *((const my_ptr2) s);
                                           ^
test.c:37:37: error: cast from 'const unsigned char *' to 'unsigned int *' drops const qualifier [-Werror,-Wcast-qual]
        *((my_ptr4) d) = *((const my_ptr4) s);
                                           ^
test.c:38:37: error: cast from 'const unsigned char *' to 'unsigned long *' drops const qualifier [-Werror,-Wcast-qual]
        *((my_ptr8) d) = *((const my_ptr8) s);
                                           ^
8 errors generated.
make: *** [Makefile:2: test] Error 1
```

As you can see, the warning only triggers when using `my_ptr{1,2,4,8}` types, not the base types, nor the `my_int{1,2,4,8}` types. As such, I believe the bug is in the interaction between the warning and the `typedef`'ing a type to a pointer type.

**Comment by gavinhoward** [comment]
This might also be related to #53889, but I didn't want to assume so.

**Comment by efriedma-quic** [comment]
I think your testcase is equivalent to the following:

```
int* f(const int* x) {
    return (int * const)x;
}
```

The "const" in the cast does nothing, so it's equivalent to just `(int*)x`, which is what the warning is supposed to warn about.
